constraints/miden-vm/chiplets.air

```diff
@@ -0,0 +1,67 @@
+use bitwise::bitwise_chiplet
+use hash::hash_chiplet
+use memory::memory_chiplet
+
+mod ChipletsConstraintsAir
+
+### Helper evaluators #############################################################################
+
+# Enforces that the provided columns must be binary.
+ev is_binary(main: [a]):
+    enf a^2 = a
+
+# Enforces that constraints are only ever selected for one chiplet at a time.
+ev chiplet_constraints(main: [s[3], chip_columns[15]]):
+    # Variables for hash chiplet.
+    let hash_s_3 = [s[1], s[2], chip_columns[0]]
+    let hash_r = chip_columns[1]
+    let hash_h_12 = [chip_columns[2], chip_columns[3], chip_columns[4], 
+                     chip_columns[5], chip_columns[6], chip_columns[7], 
+                     chip_columns[8], chip_columns[9], chip_columns[10], 
+                     chip_columns[11], chip_columns[12], chip_columns[13]]
+    let hash_i = chip_columns[14]
+
+    # Variables for bitwise chiplet.
+    let bitwise_s = s[2]
+    let bitwise_a = chip_columns[0]
+    let bitwise_b = chip_columns[1]
+    let bitwise_a_limb_4 = [chip_columns[2], chip_columns[3], chip_columns[4], chip_columns[5]]
+    let bitwise_b_limb_4 = [chip_columns[6], chip_columns[7], chip_columns[8], chip_columns[9]]
+    let bitwise_zp = chip_columns[10]
+    let bitwise_z = chip_columns[11]
+
+    # Variables for memory chiplet.
+    let memory_s_2 = [chip_columns[0], chip_columns[1]]
+    let memory_ctx = chip_columns[2]
+    let memory_addr = chip_columns[3]
+    let memory_clk = chip_columns[4]
+    let memory_v_4 = [chip_columns[5], chip_columns[6], chip_columns[7], chip_columns[8]]
+    let memory_d_2 = [chip_columns[9], chip_columns[10]]
+    let memory_t = chip_columns[11]
+
+    match enf: 
+        hash_chiplet([hash_s_3, hash_r, hash_h_12, hash_i]) when !s[0]
+        bitwise_chiplet([bitwise_s, bitwise_a, bitwise_b, bitwise_a_limb_4, bitwise_b_limb_4, bitwise_zp, bitwise_z]) when s[0] & !s[1]
+        memory_chiplet([memory_s_2, memory_ctx, memory_addr, memory_clk, memory_v_4, memory_d_2, memory_t]) when s[0] & s[1] & !s[2]'
```

Comment:
We should be able to do something much simpler here. Instead of defining all these variables, we can just do the following:

```
match enf:
    hash_chiplet([s[1], s[2], chip_columns]) when 1s[0]
    bitwise_chiplet([s[2], chip_columns]) when s[0] & !s[1]
    memory_chiplet([chip_columns]) when s[0] & s[1] & !s[2]'
```

Reply:
You mean that we will implement feature for evaluators so they would be able to rearrange input values?
So for `hash_chiplet` evaluator, for example, we could pass `[scalar, scalar, array[15]]` and evaluator will internally rearrange it to the `[array[3], scalar, array[12], scalar]`, which is declared in the evaluator definition, right? With the only condition that the input data lengths must match, of course.

Reply:
The evaluator doesn't rearrange anything really. When it is first processed, every constraint is added to the graph using trace accesses that are offset from zero according to how the parameters were defined. When an evaluator function call is processed, every SymbolAccess that is passed as an argument is processed into 1 or more trace accesses depending on how many columns it represents. This means that the number of columns in each segment should match between the defined parameters and the passed arguments, but the "shape" of these column groupings does not matter. Here's the [relevant code snippet](https://github.com/0xPolygonMiden/air-script/blob/next/ir/src/constraint_builder/evaluators.rs#L93).